File: java-source/src/main/java/net/corda/training/state/IOUState.java
package net.corda.training.state;

import com.google.common.collect.ImmutableList;
import java.util.*;

import net.corda.core.contracts.Amount;
import net.corda.core.contracts.LinearState;
import net.corda.core.contracts.UniqueIdentifier;
import net.corda.core.identity.Party;
import net.corda.core.identity.AbstractParty;
import net.corda.core.serialization.ConstructorForDeserialization;

import net.corda.training.contract.IOUContract;

/**
 * The IOU State object, with the following properties:
 * - [amount] The amount owed by the [borrower] to the [lender]
 * - [lender] The lending party.
 * - [borrower] The borrowing party.
 * - [paid] Records how much of the [amount] has been paid.
 * - [linearId] A unique id shared by all LinearState states representing the same agreement throughout history within
 *   the vaults of all parties. Verify methods should check that one input and one output share the id in a transaction,
 *   except at issuance/termination.
 */
public class IOUState implements LinearState {
    private final Amount<Currency> amount;
    private final Party lender;
    private final Party borrower;
    private final Amount<Currency> paid;
    private final UniqueIdentifier linearId;


    // Private copy constructor
    @ConstructorForDeserialization 
    private IOUState(Amount<Currency> amount, Party lender, Party borrower, Amount<Currency> paid, UniqueIdentifier linearId) {
        this.amount = amount;
        this.lender = lender;
        this.borrower = borrower;
        this.paid = paid;
        this.linearId = linearId;
    }

    // For new states
    public IOUState(Amount<Currency> amount, Party lender, Party borrower) {
        this(amount, lender, borrower, new Amount<>(0, amount.getToken()), new UniqueIdentifier());
    }

    public Amount<Currency> getAmount() {
        return amount;
    }

    public Party getLender() {
        return lender;
    }

    public Party getBorrower() {
        return borrower;
    }

    public Amount<Currency> getPaid() {
    	return paid;
    }

    @Override
    public List<AbstractParty> getParticipants() {
        return ImmutableList.of(lender, borrower);
    }

    @Override
    public UniqueIdentifier getLinearId(){
    	return linearId;
    }

	/**
	 * Helper methods for when building transactions for settling and transferring IOUs.
	 * - [pay] adds an amount to the paid property. It does no validation.
	 * - [withNewLender] creates a copy of the current state with a newly specified lender. For use when transferring.
	 */
	public IOUState pay(Amount<Currency> amountToPay) {
		Amount<Currency> newAmountPaid = this.paid.plus(amountToPay);
		return new IOUState(amount, lender, borrower, newAmountPaid, linearId);
	}

	public IOUState withNewLender(Party newLender) {
		return new IOUState(amount, newLender, borrower, paid, linearId);
	}	
}

File: src/main/kotlin/net/corda/training/state/IOUState.kt
package net.corda.training.state

import net.corda.core.contracts.Amount
import net.corda.core.contracts.LinearState
import net.corda.core.contracts.UniqueIdentifier
import net.corda.core.crypto.keys
import net.corda.core.identity.AbstractParty
import net.corda.core.identity.Party
import net.corda.training.contract.IOUContract
import java.security.PublicKey
import java.util.*

/**
 * The IOU State object, with the following properties:
 * - [amount] The amount owed by the [borrower] to the [lender]
 * - [lender] The lending party.
 * - [borrower] The borrowing party.
 * - [contract] Holds a reference to the [IOUContract]
 * - [paid] Records how much of the [amount] has been paid.
 * - [linearId] A unique id shared by all LinearState states representing the same agreement throughout history within
 *   the vaults of all parties. Verify methods should check that one input and one output share the id in a transaction,
 *   except at issuance/termination.
 */
data class IOUState(val amount: Amount<Currency>,
               val lender: Party,
               val borrower: Party,
               val paid: Amount<Currency> = Amount(0, amount.token),
               override val linearId: UniqueIdentifier = UniqueIdentifier()): LinearState {
    /**
     * This function determines if the [IOUState] is relevant to a Corda node based on whether the public keys
     * of the lender or borrower are known to the node, i.e. if the node is the lender or borrower.
     *
     * We do this by checking that the set intersection of the vault public keys with the participant public keys
     * is not the empty set.
     */
    override fun isRelevant(ourKeys: Set<PublicKey>): Boolean {
        val participantKeys = participants.map { it.owningKey }
        return ourKeys.intersect(participantKeys).isNotEmpty()
    }

    /**
     *  This property holds a list of the nodes which can "use" this state in a valid transaction. In this case, the
     *  lender or the borrower.
     */
    override val participants get() = listOf(lender, borrower)

    /**
     * A toString() helper method for displaying IOUs in the console.
     */
    override fun toString() = "IOU($linearId): ${borrower.name} owes ${lender.name} $amount and has paid $paid so far."

    /**
     * A reference to the IOUContract. Make sure this is not part of the [IOUState] constructor, if it is then
     * equality won't work property on this state type. ** Don't change this property! **
     */
    override val contract get() = IOUContract()

    /**
     * Helper methods for when building transactions for settling and transferring IOUs.
     * - [pay] adds an amount to the paid property. It does no validation.
     * - [withNewLender] creates a copy of the current state with a newly specified lender. For use when transferring.
     */
    fun pay(amountToPay: Amount<Currency>) = copy(paid = paid.plus(amountToPay))
    fun withNewLender(newLender: Party) = copy(lender = newLender)
}